c++/src/kj/refcount.h

```diff
@@ -115,6 +142,105 @@ Own<T> Refcounted::addRefInternal(T* object) {
   return Own<T>(object, *refcounted);
 }
 
+template <typename T>
+Rc<T> Refcounted::addRcRefInternal(T* object) {
+  Refcounted* refcounted = object;
+  ++refcounted->refcount;
+  return Rc<T>(object);
+}
+
+template<typename T>
+class Rc: private Own<T> {
+  // Smart pointer for reference counted objects. 
+  //
+  // There are only two ways to obtain new Rc instances:
+  // - use kj::rc<T>(...) function to create new T.
+  // - use kj::Rc::addRef() and the existing Rc instance.
+  // - use EnableAddRefToThis to allow T instance to add new references to itself.
+  //
+  // kj::Rc<T> can be used everywhere Own<T> is expected.
+  //
+  // The suggested usage patterns are:
+  // - return kj::Rc as value from factory functions: 
+  //     kj::Rc<MyService> createMyService();
+  // - pass kj::Rc as rvalue to functions that need to extend T's lifetime:
+  //     void setMyService(kj::Rc<MyService>&& service)
+  // - store kj::Rc as data member:
+  //     struct MyComputation { kj::Rc<MyService> service; };
+  //
+
+public:
+  KJ_DISALLOW_COPY(Rc);
+  Rc() { }
+
+  inline Rc(Rc&& other) noexcept: Own<T>(kj::mv(other)) { }
+
+  template <typename U, typename = EnableIf<canConvert<U*, T*>()>>
+  inline Rc(Rc<U>&& other) noexcept : Own<T>(kj::mv(other)) { }
+
+
+  kj::Rc<T> addRef() {
+    T* refcounted = this->get();
+    if (refcounted != nullptr) {
+      return Refcounted::addRcRefInternal(refcounted);
+    } else {
+      return kj::Rc<T>();
+    }
+  }
+
+  Rc& operator=(decltype(nullptr)) {
+    Own<T>::operator=(nullptr);
+    return *this;
+  }
+
+  Rc& operator=(Rc&& other) {
+    Own<T>::operator=(kj::mv(other));
+    return *this;
+  }
+
+  template <typename U>
+  Rc<U> downcast() {
+    auto own = Own<T>::template downcast<U>();
+    return Rc<U>(kj::mv(own));
+  }
+
+  inline bool operator==(decltype(nullptr)) { return this->get() != nullptr; }
```

Comment:
Non-blocking for now... since we're not exposing the `T*`, implementing `operator==(const Rc<T>&)` would be helpful.